File: src/main/java/com/yupi/airouter/service/UserProviderKeyService.java
package com.yupi.airouter.service;

import com.mybatisflex.core.service.IService;
import com.yupi.airouter.model.dto.byok.UserProviderKeyAddRequest;
import com.yupi.airouter.model.dto.byok.UserProviderKeyUpdateRequest;
import com.yupi.airouter.model.entity.UserProviderKey;
import com.yupi.airouter.model.vo.UserProviderKeyVO;

import java.util.List;

/**
 * 用户提供者密钥服务
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
public interface UserProviderKeyService extends IService<UserProviderKey> {

    /**
     * 添加用户提供者密钥
     *
     * @param request 添加请求
     * @param userId  用户 ID
     * @return 是否成功
     */
    boolean addUserProviderKey(UserProviderKeyAddRequest request, Long userId);

    /**
     * 更新用户提供者密钥
     *
     * @param request 更新请求
     * @param userId  用户 ID
     * @return 是否成功
     */
    boolean updateUserProviderKey(UserProviderKeyUpdateRequest request, Long userId);

    /**
     * 删除用户提供者密钥
     *
     * @param id     密钥 ID
     * @param userId 用户 ID
     * @return 是否成功
     */
    boolean deleteUserProviderKey(Long id, Long userId);

    /**
     * 获取用户的所有提供者密钥列表
     *
     * @param userId 用户 ID
     * @return 密钥列表
     */
    List<UserProviderKeyVO> listUserProviderKeys(Long userId);

    /**
     * 获取用户在指定提供者的密钥（解密后）
     *
     * @param userId     用户 ID
     * @param providerId 提供者 ID
     * @return 解密后的 API Key，如果不存在返回 null
     */
    String getUserProviderApiKey(Long userId, Long providerId);

    /**
     * 检查用户是否配置了指定提供者的密钥
     *
     * @param userId     用户 ID
     * @param providerId 提供者 ID
     * @return 是否配置
     */
    boolean hasUserProviderKey(Long userId, Long providerId);
}

File: python-backend/app/services/user_provider_key_service.py
"""User provider key service (BYOK)."""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.constants import BYOK_STATUS_ACTIVE, ErrorCode
from app.exceptions.business_exception import BusinessException
from app.models.model_provider import ModelProvider
from app.models.user_provider_key import UserProviderKey
from app.schemas.byok import UserProviderKeyAddRequest, UserProviderKeyUpdateRequest, UserProviderKeyVO
from app.utils.encryption import decrypt_text, encrypt_text


class UserProviderKeyService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def add_user_provider_key(self, request: UserProviderKeyAddRequest, user_id: int) -> bool:
        if request.provider_id is None or not request.api_key:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "请求参数错误")

        provider = await self.db.scalar(
            select(ModelProvider).where(
                ModelProvider.id == request.provider_id,
                ModelProvider.is_delete == 0,
            )
        )
        if provider is None:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "提供者不存在")

        exists = await self.db.scalar(
            select(UserProviderKey).where(
                UserProviderKey.user_id == user_id,
                UserProviderKey.provider_id == request.provider_id,
                UserProviderKey.is_delete == 0,
            )
        )
        if exists is not None:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "已配置过该提供者的密钥，请使用更新功能")

        entity = UserProviderKey(
            user_id=user_id,
            provider_id=request.provider_id,
            provider_name=provider.provider_name,
            api_key=encrypt_text(request.api_key),
            status=BYOK_STATUS_ACTIVE,
        )
        self.db.add(entity)
        await self.db.commit()
        return True

    async def update_user_provider_key(self, request: UserProviderKeyUpdateRequest, user_id: int) -> bool:
        entity = await self.db.scalar(
            select(UserProviderKey).where(
                UserProviderKey.id == request.id,
                UserProviderKey.is_delete == 0,
            )
        )
        if entity is None:
            raise BusinessException(ErrorCode.NOT_FOUND_ERROR, "请求数据不存在")
        if entity.user_id != user_id:
            raise BusinessException(ErrorCode.NO_AUTH_ERROR, "无权限")

        if request.api_key:
            entity.api_key = encrypt_text(request.api_key)
        if request.status:
            entity.status = request.status
        await self.db.commit()
        return True

    async def delete_user_provider_key(self, key_id: int, user_id: int) -> bool:
        entity = await self.db.scalar(
            select(UserProviderKey).where(
                UserProviderKey.id == key_id,
                UserProviderKey.is_delete == 0,
            )
        )
        if entity is None:
            raise BusinessException(ErrorCode.NOT_FOUND_ERROR, "请求数据不存在")
        if entity.user_id != user_id:
            raise BusinessException(ErrorCode.NO_AUTH_ERROR, "无权限")

        entity.is_delete = 1
        await self.db.commit()
        return True

    async def list_user_provider_keys(self, user_id: int) -> list[UserProviderKeyVO]:
        rows = (
            await self.db.scalars(
                select(UserProviderKey)
                .where(
                    UserProviderKey.user_id == user_id,
                    UserProviderKey.is_delete == 0,
                )
                .order_by(UserProviderKey.create_time.desc())
            )
        ).all()
        return [self.to_vo(row) for row in rows]

    async def get_user_provider_api_key(self, user_id: int, provider_id: int) -> str | None:
        entity = await self.db.scalar(
            select(UserProviderKey).where(
                UserProviderKey.user_id == user_id,
                UserProviderKey.provider_id == provider_id,
                UserProviderKey.status == BYOK_STATUS_ACTIVE,
                UserProviderKey.is_delete == 0,
            )
        )
        if entity is None:
            return None
        try:
            return decrypt_text(entity.api_key)
        except Exception:
            return None

    async def has_user_provider_key(self, user_id: int, provider_id: int) -> bool:
        return await self.get_user_provider_api_key(user_id, provider_id) is not None

    @staticmethod
    def to_vo(entity: UserProviderKey) -> UserProviderKeyVO:
        masked = "****"
        if entity.api_key and len(entity.api_key) > 12:
            masked = f"{entity.api_key[:8]}****{entity.api_key[-4:]}"
        return UserProviderKeyVO(
            id=entity.id,
            providerId=entity.provider_id,
            providerName=entity.provider_name,
            apiKey=masked,
            status=entity.status,
            createTime=entity.create_time,
            updateTime=entity.update_time,
        )
